src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.h

```diff
@@ -0,0 +1,47 @@
+/*******************************************************************\
+
+Module: Dynamic frame condition checking
+
+Author: Remi Delmas, delmasrd@amazon.com
+
+Date: March 2023
+
+\*******************************************************************/
+
+/// \file
+/// Utility functions that compute root object expressions for assigns clause
+/// targets and LHS expressions.
+
+#ifndef CPROVER_GOTO_INSTRUMENT_CONTRACTS_DYNAMIC_FRAMES_DFCC_ROOT_OBJECT_H
+#define CPROVER_GOTO_INSTRUMENT_CONTRACTS_DYNAMIC_FRAMES_DFCC_ROOT_OBJECT_H
+
+#include <util/irep.h>
+
+#include <unordered_set>
+
+class exprt;
+class namespacet;
+class messaget;
+
+/// Computes a set of root object expressions for an assignment left hand side
+/// expression.
+std::unordered_set<exprt, irep_hash> dfcc_assignment_lhs_root_objects(
+  const exprt &lhs,
+  const namespacet &ns,
+  messaget &log);
+
+/// Computes a set of root object expressions for an assigns clause target
+/// expression that may contain disjunctions.
+std::unordered_set<exprt, irep_hash> dfcc_assigns_target_root_objects(
+  const exprt &expr,
+  const namespacet &ns,
+  messaget &log);
+
+/// Computes a single root object expressions for an assigns clause target that
+/// is expected not to contain disjunctions. Fails if the expression contains
+/// disjunctions.
+exprt dfcc_assigns_target_single_root_object(
+  const exprt &expr,
+  const namespacet &ns,
+  messaget &log);
```

Comment:
Moved 2/3 functions elsewhere

Reply:
The documentation of the methods is here for that.

Reply:
I removed one of them. Otherwise the method name and documentation should make it clear enough. These are really custom methods for frame condition instrumentation, hence the dfcc prefix

Reply:
removed 2/3 methods, only one left.